platforms/core-runtime/service-registry-impl/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -461,8 +460,8 @@ public void add(SingletonService serviceProvider) {
             assertMutable();
             stoppable.add(serviceProvider);
             collectProvidersForClassHierarchy(inspector, serviceProvider.getDeclaredServiceTypes(), serviceProvider);
```

Comment:
❓ The change fixes the services list update, but the `providersByType` update remains unsynchronized.